rfcs/unboxed-types.md

```diff
@@ -0,0 +1,1989 @@
+
+1. We have a type signature for `g`, but it mentions a unification
+variable `'a`. We choose `'a : any`.
+
+2. Processing the body of `g` doesn't yield anything interesting,
+and so we assign `g` the type `('a : any). unit -> 'a -> 'a`,
+which is strictly more general than the type `g` is assigned today.
+
+### Layout checking
+
+Thanks to how types are constructed in OCaml, we need no separate
+well-formedness check. Suppose we're given the following type:
+
+```ocaml
+type ('a : immediate) t = ...
+```
+
+and the user attempts to write `string t`. The syntax `string t` is
+converted to a type expression by first constructing a template with a
+fresh unification variable `'b t`, and then unifying `'b` with
+`string`. Here, since `'b` ranges over layout `immediate`, the
+unification with `string` will fail.
+
+# Extensions and other addenda
+
+## Alternative idea: `#` as an operator
+
+One alternative idea we considered was to make `#` a (partial)
+operator on the type algebra, computing an unboxed version of
+a type, given its declaration. This ran aground fairly quickly,
+though.
+
+One problem is that OCaml doesn't have the ability to extract
+a type argument from a composite type (like to get the `int`
+from `int list`). So viewing `#` as the inverse of `box` doesn't
+quite work. (Richard thought for a little bit that
+`type 'a unlist = 'b constraint 'a = 'b list` did this unwrapping,
+but it doesn't: it just requires that the argument to `unlist` be
+statically a `list` type; this is not powerful enough for our needs,
+because it would reject `'a. ... # 'a ...`, which would be part
+of this more general operator's purview.)
+
+Another problem is that `#` would have to have an elaborate type:
+it would need to be something like `# : Pi (t : value) -> t layout_of_unboxed_version_of`,
+necessitating the definition of `layout_of_unboxed_version_of`.
+This is a dependent type. Dependent types are cool and all, but
+unboxed types seem like a poor motivation for them.
+
+## Extension: Restrictions around unexpected copy operations
+
+Today, OCaml users expect a line like `let x = y` to be quick, copying a word of
+data and requiring ~1 instruction. However, if `y` is a wide unboxed record,
+this assignment might take many individual copy operations and be unexpectedly
+slow. The same problem exists when an unboxed record is passed as a function
+argument. Perhaps worse, this kind of copy can happen invisibly, as part of
+closure capture. Example:
+
+```ocaml
+type t = { a : int; b : int; c : int; d : int; e : int }
+
+let f (t : #t) =
+  let g n = frob (n+1) t in
+  ...
+```
+
+Building the closure for `g` requires copying the five fields of `t`.
+
+We could imagine a restriction in the language stopping such copies from
+happening, or requiring extra syntax to signal them. However, we hope that there
+will not be too many wide unboxed records, and copying a few fields really isn't
+so slow. So, on balance, we do not think this is necessary.
```

Comment:
Yes. There has been considerable discomfort around this point internally as well. But putting e.g. `#<-` on every mutation seems pretty unpleasant, too. I really don't know what's best.

Reply:
A comparison with C++ is in fact somewhat unfair, as the silent copies could be recursive (e.g. copying a vector of vectors) so the problem is much worse. A better comparison is with Rust, so it can be interesting to read what they say about it (e.g. https://github.com/rust-lang/rust/issues/45683).

Rust has unboxed fixed-size arrays, unlike this proposal, which make it easier to make copies costly. However, I argue that this difference is not important. Indeed, fixed-size arrays are useful, and with your proposal, users can emulate them by writing them explicitly with something like `t * t * ... * t`, so I believe that code like this with large structs are going to be written regardless.

Reply:
My sense reading that thread is that they are just about as puzzled as we are here.

In my opinion, we should not add extra annotations to the language for this. Instead, the goal we are after is this: we want to make sure the programmer knows what they are doing. I can think of at least three ways of accomplishing this goal:

1. Add required annotations to the language
2. Add a warning; we can then debate whether the warning should be on by default. And we have to figure out a way of locally suppressing the warning, likely with some kind of annotation.
3. Add magic to merlin to alert the user.

I'm strongly in favor of 3, if it can be made to work. That is, imagine if an expensive `<-` is highlighted in yellow, say. The programmer can see this and account for it. Maybe we also add an annotation to suppress the highlighting? Not sure there. It would be very cool if an editor with merlin enabled rendered `#<-` as `<-` with a little `#` over it, reducing visual clutter. Actually, maybe this is the same as (2), but with special treatment in merlin.

To be clear, my "if it can be made to work" is a hedge against the possibility that people don't actually use merlin. My OCaml experience is very skewed: I'm in an environment where everybody uses merlin, because it's all set up for them. I'm not sure it's as pervasive outside Jane Street's walls.

Reply:
It sounds to me like you are advocating for a lint, like in this Rust thread.

I do not have a definite opinion or an alternative solution to propose, but tooling is less mature in OCaml and has fewer contributors, and if the problem is deemed worthy of language design considerations then one should be careful about not relegating the problem to future/incomplete/third-party support.